apps/meteor/app/push/server/fcm.ts

```diff
@@ -0,0 +1,159 @@
+type FCMError = {
+	error: {
+		code: number;
+		message: string;
+		status: string;
+	};
+};
+
+function fetchWithRetry(url: string, options: RequestInit, retries = 0): Promise<Response> {
+	const MAX_RETRIES = 5;
+	const ERRORS_TO_RETRY = [429, 500, 502, 503, 504];
+
+	try {
+		return fetch(url, options);
+	} catch (error) {
+		if (retries >= MAX_RETRIES) {
+			throw error;
+		}
+
+		const errorData = error as FCMError;
+
+		if (!errorData?.error?.code || !ERRORS_TO_RETRY.includes(errorData.error.code)) {
+			throw error;
+		}
+
+		return fetchWithRetry(url, options, retries + 1);
+	}
+}
+
+function getFCMMessagesFromPushData(userTokens: string[], notification: PendingPushNotification): { message: FCMMessage }[] {
+	// first we will get the `data` field from the notification
+	const data: FCMDataField = notification.payload ? { ejson: EJSON.stringify(notification.payload) } : {};
+
+	// Set image
+	if (notification.gcm?.image != null) {
+		data.image = notification.gcm?.image;
+	}
+
+	// Set extra details
+	if (notification.badge != null) {
+		data.msgcnt = notification.badge.toString();
+	}
+	if (notification.sound != null) {
+		data.soundname = notification.sound;
+	}
+	if (notification.notId != null) {
+		data.notId = notification.notId.toString();
+	}
+	if (notification.gcm?.style != null) {
+		data.style = notification.gcm?.style;
+	}
+
+	if (notification.contentAvailable != null) {
+		data['content-available'] = notification.contentAvailable.toString();
+	}
+
+	// then we will create the notification field
+	const notificationField: FCMNotificationField = {
+		title: notification.title,
+		body: notification.text,
+	};
+
+	// then we will create the message
+	const message: FCMMessage = {
+		notification: notificationField,
+		data,
+		android: {
+			priority: 'HIGH',
+		},
+	};
+
+	// then we will create the message for each token
+	return userTokens.map((token) => ({ message: { ...message, token } }));
+}
+
+export const sendFCM = function ({ userTokens, notification, _replaceToken, _removeToken, options }: NativeNotificationParameters): void {
+	// We don't use these parameters, but we need to keep them to keep the function signature
+	// TODO: Remove them when we remove the old sendGCM function
+	_replaceToken;
+	_removeToken;
+
+	const tokens = typeof userTokens === 'string' ? [userTokens] : userTokens;
+	if (!tokens.length) {
+		logger.debug('sendFCM no push tokens found');
+		return;
+	}
+
+	logger.debug('sendFCM', tokens, notification);
+
+	const messages = getFCMMessagesFromPushData(tokens, notification);
+	const headers = {
+		'Content-Type': 'application/json',
+		'Authorization': `Bearer ${options.gcm.apiKey}`,
+		'access_token_auth': true,
+	} as Record<string, any>;
+
+	const credentialsString = settings.get<string>('Push_google_api_credentials');
+	const credentials = JSON.parse(credentialsString) as FCMCredentials;
+
+	const url = `https://fcm.googleapis.com/v1/projects/${credentials.project_id}/messages:send`;
```

Comment:
it will pass ajv validation even if `project_id` is an empty string afaik. with retry being in place, would be better to validate first that it isn't an empty string before starting that story.